File: src/tmp/Python/refer-offer/zui_chang_bu_han_zhong_fu_zi_fu_de_zi_zi_fu_chuan_lcof.py
import collections

def lengthOfLongestSubstring(s: str) -> int:
    left, right, length, res = 0, 0, len(s), 0
    remain = collections.Counter()
    while right < length:
        if remain[s[right]] != 0:
            while remain[s[right]] != 0:
                remain[s[left]] -= 1
                left += 1
        
        remain[s[right]] += 1
        res = max(res, right - left + 1)
        right += 1
    
    return res

File: src/tmp/Java/src/com/dbc/algorithm/ReferOffer/ZuiChangBuHanZhongFuZiFuDeZiZiFuChuanLcof.java
package com.dbc.algorithm.ReferOffer;

import java.util.HashMap;
import java.util.Map;

public class ZuiChangBuHanZhongFuZiFuDeZiZiFuChuanLcof {
    public int lengthOfLongestSubstring(String s) {
        int left = 0, right = 0, res = 0;
        Map<Character, Integer> map = new HashMap<>();
        while (right < s.length()) {
            if (map.getOrDefault(s.charAt(right), 0) != 0) {
                while (map.getOrDefault(s.charAt(right), 0) != 0) {
                    map.put(s.charAt(left), map.get(s.charAt(left)) - 1);
                    left++;
                }
            }
            map.put(s.charAt(right), map.getOrDefault(s.charAt(right), 0) + 1);
            res = Math.max(res, right - left + 1);
            right++;
        }
        return res;
    }
}
